cores/esp8266/core_esp8266_spi_utils.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ core_esp8266_spi_utils.cpp
+
+ Copyright (c) 2019 Mike Nix. All rights reserved.
+ This file is part of the esp8266 core for Arduino environment.
+
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2.1 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include <stdint.h>
+#include <string.h>
+
+// register names
+#include "esp8266_peri.h"
+
+// for flashchip
+#include "spi_flash.h"
+
+// for PRECACHE_*
+#include "core_esp8266_features.h"
+
+#include "spi_utils.h"
+
+extern "C" uint32_t Wait_SPI_Idle(SpiFlashChip *fc);
+
+/*
+ * critical part of SPICommand.
+ * Kept in a separate function to aid with precaching
+ * PRECACHE_* saves having to make the function IRAM_ATTR.
+ *
+ * Note: if porting to ESP32 mosi/miso bits are set in 2 registers, not 1.
+ */
+static SpiOpResult PRECACHE_ATTR
+_SPICommand(volatile uint32_t spiIfNum,
+            uint32_t spic,uint32_t spiu,uint32_t spiu1,uint32_t spiu2,
+            uint32_t *data,uint32_t write_words,uint32_t read_words)
+{ 
+  if (spiIfNum>1) return SPI_RESULT_ERR;
+
+  // force SPI register access via base+offest. 
+  // Prevents loading individual address constants from flash.
+  uint32_t *spibase = (uint32_t*)(spiIfNum ? &(SPI1CMD) : &(SPI0CMD));
```

Comment:
spibase doesn't need to be volatile?

Reply:
made no difference to the compiled code when it was :-)